### PR TITLE
GH-39579: [Python] fix raising ValueError on _ensure_partitioning

### LINKED ISSUE #39579: exception not being raised if passed an invalid partitiong on _ensure_partitioning
### Describe the bug, including details regarding any error messages, version, and platform.

It seems to be a typo at line 295 of python/pyarrow/dataset.py, which misses a "raise"

```
def _ensure_partitioning(scheme):
    ...
    else:
        ValueError(...)

# which should be
    else:
        raise ValueError(...)
```

### Component(s)

Python

### PR DESCRIPTION
**Pull request by idailylife**
### Rationale for this change
The `_ensure_partitioning` method in dataset.py is missing a "raise" which currently ignores bad scheme silently.

### What changes are included in this PR?

Fixed the typo.

### Are these changes tested?
Tried with new code that the exception is properly raised.

### Are there any user-facing changes?
No.

* Closes: #39579

### CODE REVIEW

**Comment by AlenkaF** [comment]
Thank you for catching this!
We should add a test for it though, preferably in `pyarrow/tests/dataset.py::test_partitioning`.

**Comment by AlenkaF** [comment]
Can we add a test to `pyarrow/tests/dataset.py::test_partitioning`?

**Comment by idailylife** [comment]
> Can we add a test to `pyarrow/tests/dataset.py::test_partitioning`?

Yup working on it

**Comment by idailylife** [comment]
Not sure why one of the CIs failed, seemingly unrelated.

File: python/pyarrow/tests/test_dataset.py [hunk 1]
```diff
@@ -701,6 +701,16 @@ def test_partitioning():
             load_back_table = load_back.to_table()
             assert load_back_table.equals(table)
 
+    # test invalid partitioning input
+    with tempfile.TemporaryDirectory() as tempdir:
+        partitioning = ds.DirectoryPartitioning(partitioning_schema)
+        ds.write_dataset(table, tempdir,
+                         format='ipc', partitioning=partitioning)
+        load_back = None
+        with pytest.raises(ValueError):
```
Comment (AlenkaF) [inline python/pyarrow/tests/test_dataset.py:710]:
```suggestion
        with pytest.raises(ValueError,
                           match="Expected Partitioning or PartitioningFactory"):
```

**Comment by AlenkaF** [comment]
The failure is not connected, I can see it elsewhere also: https://github.com/ursacomputing/crossbow/actions/runs/7601318025/job/20700571533

**Review by AlenkaF — APPROVED** [review]
The failures are not connected and have an issue opened: https://github.com/apache/arrow/issues/39732

Thanks for the fix!

**Pull request merged**